frontend/public/components/storage/attach-storage.tsx

```diff
@@ -0,0 +1,437 @@
+    />
+  );
+};
+
+const StorageForm = () =>
+  class StorageFormComponent extends React.Component<
+    StorageProps,
+    StorageState
+  > {
+    constructor(props) {
+      super(props);
+      const existingResource = _.pick(props.obj, ["metadata", "spec"]);
+      const resourceObj = _.defaultsDeep({}, props.fixed, existingResource, {
+        apiVersion: "v1",
+        data: {},
+        kind: props.kind,
+        metadata: {
+          name: ""
+        },
+        spec: {}
+      });
+
+      this.state = {
+        resourceObj: resourceObj,
+        inProgress: false,
+        stringData: _.mapValues(_.get(props.obj, "data"), value => {
+          return value ? window.atob(value) : "";
+        }),
+        disableForm: false
+      };
+      this.onVolumeNameChanged = this.onVolumeNameChanged.bind(this);
+      this.onMountPathChanged = this.onMountPathChanged.bind(this);
+      this.onSubPathChanged = this.onSubPathChanged.bind(this);
+      this.changePvcName = this.changePvcName.bind(this);
+      this.onReadOnlyChanged = this.onReadOnlyChanged.bind(this);
+      this.onError = this.onError.bind(this);
+      this.onFormDisable = this.onFormDisable.bind(this);
+      this.save = this.save.bind(this);
+    }
+    componentDidMount() {
+      const { kind, metadata } = this.state.resourceObj;
+      const ko = _.find(k8sModels, model => model.plural === kind);
+      k8sGet(ko, metadata.name, metadata.namespace).then(
+        obj => {
+          this.setState({ resourceObj: obj });
+        },
+        err => this.setState({ error: err.message, inProgress: false })
+      );
+    }
+    onError(err) {
+      this.setState({
+        error: err,
+        inProgress: false
+      });
+    }
+    changePvcName(event) {
+      let stringData = { ...this.state.stringData };
+      stringData.claimName = event;
+      this.setState({ stringData });
+    }
+    onVolumeNameChanged(event) {
+      let stringData = { ...this.state.stringData };
+      stringData.volumeName = event.target.value;
+      this.setState({ stringData });
+    }
+    onMountPathChanged(event) {
+      let stringData = { ...this.state.stringData };
+      stringData.mountPath = event.target.value;
+      this.setState({ stringData });
+    }
+    onSubPathChanged(event) {
+      let stringData = { ...this.state.stringData };
+      stringData.subPath = event.target.value;
+      this.setState({ stringData });
+    }
+    onReadOnlyChanged(event) {
+      let stringData = { ...this.state.stringData };
+      stringData.readOnly = (stringData.readOnly === "read-only") ? "read-write" : "read-only";
+      this.setState({ stringData });
+    }
+    onFormDisable(disable) {
+      this.setState({
+        disableForm: disable
+      });
+    }
+    // Add Attach storage //
+    save(e) {
+      const { kind, metadata } = this.state.resourceObj;
+      const ko = _.find(k8sModels, model => model.kind === kind);
+      let resourceObj = { ...this.state.resourceObj };
```

Comment:
const

Reply:
This is only doing a shallow copy of resource obj, and we're changing nested fields.

We should try to use patch instead of put to avoid conflicts.

Reply:
I cleaned this up, but not sure what you mean "We should try to use patch instead of put to avoid conflicts."